2-Draft-Accepted/RFCxxxx-PowerShellGet-3.0.md

```diff
@@ -0,0 +1,383 @@
+the parameter names (like splatting).
+
+```powershell
+Register-PSResourceRepository -Repositories @(
+  @{ URL = "https://one.com"; Trusted = $true }
+  @{ URL = "https://powershellgallery.com"; Trusted = $true; Default = $true }
+)
+```
+
+A `-Name` parameter allows for setting a friendly name.
+
+A `-Default` switch will set one repository as the default (when not specified
+with other cmdlets).
+Each time it is specified, it sets the new one as default and the previous default
+is no longer default.
+
+`Get-PSResourceRepository` will list out the registered repositories.
+
+`Set-PSResourceRepository` can be used to update a repository URL, trust level,
+or if that repository is the default.
+
+`Unregister-PSResourceRepository` can be used to un-register a repository.
+
+### Finding resources
+
+Current design of PowerShellGet is to have different cmdlets for each of the
+different resources types that it supports:
+
+`Find-Command`
+`Find-DscResource`
+`Find-Module`
+`Find-RoleCapability`
+`Find-Script`
+
+Instead, these will be abstracted by a `Find-PSResource` cmdlet.
+A `-Type` parameter will accept an array of types to allow filtering.
+
+With support of the generic `PSResource`, this means we can also find and
+install arbitrary nupkgs that may only contain an assembly the user wants to
+use in their script.
+
+For each repository, if the local cache does not exist, the cmdlet will call
+`Update-PSResourceCache` first and then attempt a search.
+However, this cmdlet does not update the cache if one exists.
+
+`Find-Module` will be retained to provide compatibility with v2.
+
+### Installing resources
+
+`Install-PSResource` will only work for modules unless `-DestinationPath` is
+specified which works for all resource types.
+A `-Prerelease` switch allows installing prerelease versions.
+Other types will use `Save-PSResource` (see below).
+A `-Repository` parameter accepts a specific repository name or URL to the repository:
+
+```powershell
+Install-PSResource myModule -Repository 'https://mygallery.com'
+
+# or
+
+Install-Module myModule -Repository 'https://mygallery.com'
+```
+
+If `-Repository` is not specified and the same resource is found in multiple
+repositories, then the resource is automatically installed quietly from the
+trusted repository with the highest version matching the `-Version` parameter
+(if specified, otherwise newest non-prerelease version unless `-Prerelease`
+is used).
+If there are no trusted repositories matching the query, then the newest version
+fulfilling the query will be prompted to be installed.
+If there are multiple repositories with the same trust level containing the same
+version, the first one is used.
+
+`-TrustRepository` can be used to suppress being prompted for untrusted sources.
+`-IgnoreDifferentPublisher` can be used to suppress being prompted if the publisher
+of the module is different from the currently installed version.
+`-Reinstall` can be used to re-install a module.
+
```

Comment:
I'm not a fan of `-Force`.  The intent of this RFC is those switches should be less used most of the time and should only be needed in exceptional cases.  Most users should never have to worry about them.  If you feel that isn't the case, please call it out.

As for cleaning old versions, this is explicitly called out as out of scope for this RFC/iteration.  Certainly something I'd like to address in the future.  That whole feature should be its own RFC.

Reply:
My concern is that this motivation statement seems to be in conflict with the `-TrustRepository`, -`IgnoreDifferentPublisher` and even `-Reinstall` switches:

>  As a PowerShell user,
>    I can easily install modules without specifying lots of switches,
>    so that I can be more productive.

Actually, that motivation statement should probably be expanded to the following so that it meets the goals of many existing PowerShellGet users:

As a PowerShell user,
  I can easily install modules without knowing or specifying lots of switches,
  so that I can be more productive.

Forcing users to learn individual switches to deal with different situations is wrong.

Two of those switches are each about suppression of a prompt, they are not commonly known, and require lookup to use. The other is about installing a module even if you have it installed. `-Force` is common, is well known without reading docs, and should just do the trick no matter what. Having a bunch of individual switches to apply in different situations when you just want something done is an unnecessary annoyance that should be avoided. Just let users `-Force` it when they know what they're doing and don't want to have to read docs to get it done. That's what that motivation statement should be about. Otherwise you're just replacing one annoyance (the switches in PowerShell Get 2.x) with another (the new collection of switches you need to know for different situations).